src/machine/machine_atsamd51.go

```diff
@@ -1202,6 +1270,34 @@ func (spi SPI) Transfer(w byte) (byte, error) {
 	return byte(spi.Bus.DATA.Get()), nil
 }
 
+// Transfer2 writes/reads a single uint16 using the SPI interface.
+func (spi SPI) Transfer2(w uint16) (uint16, error) {
```

Comment:
I see you have extended the SPI peripheral with 16-bit and 32-bit methods.
1. Does this provide an improvement in speed?
2. Is this really necessary for WioTerminal support?

I'm asking this, because eventually I would like to switch to DMA and interrupt based SPI transfers. That would make it unnecessary to add `Transfer2` and `Transfer4` methods, as all the traffic is handled by DMA (which is the fastest possible transfer, except maybe for very short transfers).

Reply:
Yes.
In the demo of pyportal_boing, ILI9341 is too slow at about 10 fps when using Transfer().
Using Transfer2() makes the ILI9341 32 fps and Transfer4() makes it 36 fps.

The processing time excluding DrawRGBBitmap() was 12.5ms, so the time it takes for one DrawRGBBitmap() is as follows

* Transfer : 100 ms
* Transfer2 : 31.25 ms
* Transfer3 : 27.8 ms

Also, I think that DMA should be enabled at some point.
But without a DMA, it's too slow to use Trasnfer4.
So I'd like to enable Transfer4 once, if possible.

Reply:
You may be able to use a trick like here to improve the speed:
https://github.com/tinygo-org/tinygo/pull/562
What it does, is that it makes use of double buffering support in the hardware. This makes it possible to continuously stream out data. Maybe the SAM D5x also supports double buffering in the `DATA` register.

Reply:
By using INTEFLAG.DRE, I was able to improve performance without using 32bit-extensions.
ILI9341's pyportal_boing is now 43 fps instead of 36 fps.

Reply:
> ILI9341's pyportal_boing is now 43 fps instead of 36 fps.

I think you possibly meant "36 fps instead of 43 fps" :smile_cat:

Reply:
The double-buffer-like processing of DRE checks seems to be faster than the 4-byte writes ( the 32B-extensions).

| board | cpu | clock | driver | fps |
| ------- | ---- | ----- | ------- | ---- |
| pyportal | samd51j20 | 120Mhz | parallel | 51 fps |
| wioterminal | samd51p19 | 120Mhz | spi (first commit) | 10 fps |
| wioterminal | samd51p19 | 120Mhz | spi with 32B extensions | 36 fps |
| wioterminal | samd51p19 | 120Mhz | spi with DRE check and without 32B extensions | 43 fps |

Reply:
That's a nice improvement! And it should benefit anyone using SPI on the atsamd51.
Can you make a separate PR with those changes?

Reply:
Sorry, now I see you already made the change in https://github.com/tinygo-org/drivers/pull/153.

Reply:
Maybe the case of just sending (spi.Tx(w, nil)) can be done faster in the same way.
I put that code in ILI9341, but I think it can be done by moving it from spi.go to machine_atsamd51.go (maybe atsamd21 can do it the same way).